Introduce graceful cancellation of invocations

tillrohrmann opened the pull request:
This commit introduces a new TerminationFlavor::Cancel which will
gracefully terminate an invocation. The way it works is that only
leaf operations are being terminated with a CANCELED_INVOCATION_ERROR.
Leaf operations are all completable journal entries except for calls
and inboxed invocations. For pending calls in a journal, a cancellation
will propagate the cancellation to these calls.

This commit adds support for canceling inboxed, invoked, suspended and
virtual invocations.

This fixes https://github.com/restatedev/restate/issues/986 and https://github.com/restatedev/restate/issues/943.

slinkydeveloper commented on crates/types/src/invocation.rs:
```diff
@@ -379,6 +379,38 @@ impl SpanRelation {
+/// Flavor of the termination. Can be kill (hard stop) or graceful cancel.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum TerminationFlavor {
+    // hard termination, no clean up
```
triple `///` for rustdoc, also below

tillrohrmann replied:
Will do.

slinkydeveloper commented on crates/storage-api/src/outbox_table/mod.rs:
```diff
@@ -29,8 +31,8 @@ pub enum OutboxMessage {
-    /// Kill command to send to another partition processor
-    Kill(FullInvocationId),
+    /// Terminate invocation to send to another partition processor
+    TerminateInvocation(InvocationTermination),
```
Maybe just `Termination` or `InvocationTermination` to be consistent with above variants?

tillrohrmann replied:
Will rename it into `InvocationTermination`.

slinkydeveloper commented on crates/storage-proto/proto/dev/restate/storage/v1/domain.proto:
```diff
@@ -14,6 +14,13 @@ message FullInvocationId {
 }
 
+message MaybeFullInvocationId {
+    oneof kind {
+        FullInvocationId full_invocation_id = 1;
+        bytes invocation_id = 2;
+    }
```
Premature optimization: Rather than this representation with nested messages, perhaps we could just do it as follows:

```protobuf
message MaybeFullInvocationId {
  int64 partition_key = 1;
  bytes invocation_uuid = 2;

  // If service name missing, this is InvocationId, otherwise it's FullInvocationId
  bytes service_name = 3;
  bytes service_key = 4;
}
```

Flattening the protobuf representation is generally more efficient than many nested messages.

tillrohrmann replied:
Indeed this is maybe premature optimizations since I am not sure whether we enforce that service names can not be empty atm. And even if we do (which probably makes sense), it introduces a subtle coupling that might break once we change things. Not sure whether the performance gains justify that.

slinkydeveloper commented on crates/worker/src/partition/shuffle.rs:
```diff
@@ -53,7 +55,7 @@ pub(crate) struct ShuffleInput(pub(crate) AckKind);
     Invocation(ServiceInvocation),
     Response(InvocationResponse),
-    Kill(FullInvocationId),
+    TerminateInvocation(InvocationTermination),
```
Just `Terminate`?

tillrohrmann replied:
Will rename it into `InvocationTermination`.

slinkydeveloper commented on crates/worker/src/partition/state_machine/command_interpreter.rs:
```diff
@@ -515,6 +641,91 @@ where
+    ) -> Result<bool, Error> {
+        let mut journal = state.get_journal(&full_invocation_id.service_id, journal_length);
+
+        let canceled_result = CompletionResult::Failure(
```
If there isn't a `From` impl, perhaps let's add one.

tillrohrmann replied:
Will add one.

slinkydeveloper commented on crates/worker/src/partition/state_machine/command_interpreter.rs:
```diff
@@ -515,6 +641,91 @@ where
+                        );
+                    }
+                    EnrichedEntryHeader::Awakeable { is_completed }
+                    | EnrichedEntryHeader::GetState { is_completed }
```
You can't do this, because in the protocol `GetState` doesn't have a failure variant.

slinkydeveloper commented on crates/worker/src/partition/state_machine/command_interpreter.rs:
```diff
@@ -515,6 +641,91 @@ where
+                    EnrichedEntryHeader::Awakeable { is_completed }
+                    | EnrichedEntryHeader::GetState { is_completed }
+                    | EnrichedEntryHeader::Sleep { is_completed }
+                    | EnrichedEntryHeader::PollInputStream { is_completed }
```
Same here, you can't do that.

tillrohrmann replied:
Thanks for spotting this. I will remove it.

slinkydeveloper commented on crates/worker/src/partition/state_machine/command_interpreter.rs:
```diff
@@ -100,6 +106,13 @@ pub trait StateReader {
 }
 
+/// Projected [`InvocationStatus`] for cancellation purposes.
+enum InvocationStatusProjection {
```
Maybe put this definition somewhere at the bottom of the file, or IDK. Here it's super confusing.

tillrohrmann replied:
Will do.

slinkydeveloper commented on crates/worker/src/partition/state_machine/command_interpreter.rs:
```diff
@@ -515,6 +641,91 @@ where
+                        }
+                    }
+                    header => {
+                        assert!(
+                            header.is_completed().unwrap_or(true),
+                            "All non canceled journal entries must be completed."
+                        );
```
debug_assert?

tillrohrmann replied:
I would like to keep it an assert so that it always fires if there is a violation.

tillrohrmann commented:
Thanks for the review @slinkydeveloper. I've addressed your comments. Concerning the layout of the `MaybeFullInvocationId`, I would suggest to postpone this optimization since it is not clear to me how we currently enforce that service names cannot be empty. PTAL.

tillrohrmann commented:
I just noticed that the Java SDK also does not allow a failure result for sleep entries. Instead of not allowing exceptionally completing some completable journal entries, I believe we should rather change the protocol to allow it and then update the SDKs respectively. Otherwise the cancellation feature will only work by cancelling inboxed invocations and nothing else.

tillrohrmann commented:
I've reverted fixup which removed sending a failed completion for `PollInputStream` and `GetState` entries. This PR is now depended on #1000.

tillrohrmann commented on crates/types/src/errors.rs:
```diff
@@ -322,6 +322,11 @@ pub const KILLED_INVOCATION_ERROR: InvocationError = InvocationError::new_static
 );
 
+pub const CANCELED_INVOCATION_ERROR: InvocationError = InvocationError::new_static(
+    InvocationErrorCode::User(UserErrorCode::Cancelled),
```
Good and valid point. I will add a comment to remind us of that once we want to introduce the distinction.
